quinn-proto/src/connection/streams/state.rs

```diff
@@ -20,12 +20,58 @@ use crate::{
     Dir, Side, StreamId, TransportError, VarInt, MAX_STREAM_COUNT,
 };
 
+/// Wrapper around `Recv` that facilitates reusing `Recv` instances
+#[derive(Debug)]
+pub(super) enum StreamRecv {
+    /// A `Recv` that is ready to be opened
+    Ready(Box<Recv>),
+    /// A `Recv` that has been opened
+    Open(Box<Recv>),
+}
+
+impl StreamRecv {
+    /// Returns a reference to the inner `Recv` if the stream is open
+    pub(super) fn open_recv(&self) -> Option<&Box<Recv>> {
+        match self {
+            StreamRecv::Open(r) => Some(r),
+            _ => None,
+        }
+    }
+
+    // Returns a mutable reference to the inner `Recv` if the stream is open
+    pub(super) fn open_recv_mut(&mut self) -> Option<&mut Box<Recv>> {
+        match self {
+            StreamRecv::Open(r) => Some(r),
+            _ => None,
+        }
+    }
+
+    // Returns the inner `Recv`
+    pub(super) fn into_inner(self) -> Box<Recv> {
+        match self {
+            StreamRecv::Ready(r) | StreamRecv::Open(r) => r,
+        }
+    }
+
+    // Reinitialize the stream so the inner `Recv` can be reused
+    pub(super) fn reinit(self, initial_max_data: u64) -> Self {
+        match self {
+            StreamRecv::Ready(_) => unreachable!("StreamRecv::reset on Ready"),
+            StreamRecv::Open(mut recv) => {
+                recv.reinit(initial_max_data);
+                StreamRecv::Ready(recv)
+            }
+        }
+    }
+}
+
 #[allow(unreachable_pub)] // fuzzing only
 pub struct StreamsState {
     pub(super) side: Side,
     // Set of streams that are currently open, or could be immediately opened by the peer
     pub(super) send: FxHashMap<StreamId, Option<Box<Send>>>,
-    pub(super) recv: FxHashMap<StreamId, Option<Box<Recv>>>,
+    pub(super) recv: FxHashMap<StreamId, Option<StreamRecv>>,
+    pub(super) free_recv: Vec<StreamRecv>,
```

Comment:
Having `free_recv` allows me to keep most of the internal API as is and grab spare streams in StreamState::insert. If you like the idea behind this PR, I'd actually want to get rid of free_recv (an extra allocation :P), and plumb streams directly into `ensure_remote_streams()` somehow

Reply:
> plumb streams directly into ensure_remote_streams() somehow

Be mindful that `Recv` streams may also be locally-initiated bidirectional streams.

> an extra allocation :P

A (more or less) one-time allocation doesn't matter much, though I concur that expanding on the existing pattern of pre-allocated entries in `send`/`recv` seems graceful.

Reply:
Because leaders rotate very frequently, in solana every validator gets a huge spike in connections right before its leader slots. After I'm done minimizing per-stream allocations, I'll want to minimize per-connection allocations too, although it's still unclear how upstreamable that part will be.